### ISSUE TITLE
Incorrect arguments for gcc causing header parse failure

### ISSUE THREAD

**Issue by rongcuid** [clangd/clangd#1907]
I have a CUDA project, where clangd fails to parse header files due to passing incorrect arguments to the gcc compiler:

```
I[16:24:26.795] <-- textDocument/codeAction(25)
I[16:24:26.796] --> reply:textDocument/codeAction(25) 0 ms
I[16:24:26.796] --> textDocument/clangd.fileStatus
I[16:24:28.146] <-- shutdown(26)
I[16:24:28.146] --> reply:shutdown(26) 0 ms
I[16:24:28.188] clangd version 17.0.3 (https://github.com/llvm/llvm-project 888437e1b60011b8a375dd30928ec925b448da57)
I[16:24:28.188] Features: linux+grpc
I[16:24:28.188] PID: 2708188
I[16:24:28.188] Working directory: /localdisk1/rdong3/cuda_atomic_utilization
I[16:24:28.188] argv[0]: /nfs/hen2-10/homes/rdong3/.vscode-server/data/User/globalStorage/llvm-vs-code-extensions.vscode-clangd/install/17.0.3/clangd_17.0.3/bin/clangd
I[16:24:28.188] Starting LSP over stdin/stdout
I[16:24:28.189] <-- initialize(0)
I[16:24:28.189] <-- exit
I[16:24:28.189] LSP finished, exiting with status 0
I[16:24:28.191] --> reply:initialize(0) 1 ms
I[16:24:28.192] <-- initialized
I[16:24:28.194] <-- textDocument/didOpen
I[16:24:28.194] <-- textDocument/didOpen
I[16:24:28.195] --> textDocument/publishDiagnostics
I[16:24:28.195] Loaded compilation database from /localdisk1/rdong3/cuda_atomic_utilization/.vscode/compile_commands.json
I[16:24:28.195] ASTWorker building file /localdisk1/rdong3/cuda_atomic_utilization/src/timeline/timeline.hpp version 1 with command inferred from src/timeline/timeline.cpp
[/localdisk1/rdong3/cuda_atomic_utilization]
/usr/bin/gcc -c -m64 -Ithirdparty -Isrc -I/usr/local/cuda/include -isystem /localdisk1/rdong3/cuda_atomic_utilization/build/.packages/a/apache-arrow/latest/4e0143c97b65425b855ad5fd03038b6a/include -std=c++17 -resource-dir=/localdisk1/rdong3/vscode-server/data/User/globalStorage/llvm-vs-code-extensions.vscode-clangd/install/17.0.3/clangd_17.0.3/lib/clang/17 -- /localdisk1/rdong3/cuda_atomic_utilization/src/timeline/timeline.hpp
I[16:24:28.195] ASTWorker building file /localdisk1/rdong3/cuda_atomic_utilization/src/cuda_allocation.hpp version 1 with command inferred from src/cmds/timeline_cmd.cpp
[/localdisk1/rdong3/cuda_atomic_utilization]
/usr/bin/gcc -c -m64 -Ithirdparty -Isrc -I/usr/local/cuda/include -isystem /localdisk1/rdong3/cuda_atomic_utilization/build/.packages/a/apache-arrow/latest/4e0143c97b65425b855ad5fd03038b6a/include -std=c++17 -resource-dir=/localdisk1/rdong3/vscode-server/data/User/globalStorage/llvm-vs-code-extensions.vscode-clangd/install/17.0.3/clangd_17.0.3/lib/clang/17 -- /localdisk1/rdong3/cuda_atomic_utilization/src/cuda_allocation.hpp
I[16:24:28.195] --> window/workDoneProgress/create(0)
I[16:24:28.195] Enqueueing 12 commands for indexing
I[16:24:28.196] <-- reply(0)
I[16:24:28.196] --> $/progress
I[16:24:28.196] --> $/progress
I[16:24:28.196] --> textDocument/clangd.fileStatus
I[16:24:28.196] --> textDocument/clangd.fileStatus
I[16:24:28.245] Built preamble of size 1122364 for file /localdisk1/rdong3/cuda_atomic_utilization/src/cuda_allocation.hpp version 1 in 0.05 seconds
I[16:24:28.245] --> workspace/semanticTokens/refresh(1)
I[16:24:28.246] <-- reply(1)
E[16:24:28.254] IncludeCleaner: Failed to get an entry for resolved path : No such file or directory
E[16:24:28.254] IncludeCleaner: Failed to get an entry for resolved path : No such file or directory
I[16:24:28.255] --> textDocument/publishDiagnostics
I[16:24:28.255] --> textDocument/inactiveRegions
I[16:24:28.256] --> textDocument/clangd.fileStatus
I[16:24:28.273] <-- textDocument/documentSymbol(1)
I[16:24:28.273] <-- textDocument/codeAction(2)
I[16:24:28.275] --> reply:textDocument/documentSymbol(1) 1 ms
I[16:24:28.275] --> reply:textDocument/codeAction(2) 1 ms
I[16:24:28.275] --> textDocument/clangd.fileStatus
I[16:24:28.286] <-- textDocument/documentLink(3)
I[16:24:28.286] --> reply:textDocument/documentLink(3) 0 ms
I[16:24:28.286] <-- textDocument/inlayHint(4)
I[16:24:28.286] --> reply:textDocument/inlayHint(4) 0 ms
I[16:24:28.286] --> textDocument/clangd.fileStatus
I[16:24:28.298] <-- textDocument/semanticTokens/full(5)
I[16:24:28.298] --> reply:textDocument/semanticTokens/full(5) 0 ms
I[16:24:28.298] --> textDocument/clangd.fileStatus
I[16:24:28.365] Built preamble of size 2561644 for file /localdisk1/rdong3/cuda_atomic_utilization/src/timeline/timeline.hpp version 1 in 0.17 seconds
I[16:24:28.365] --> workspace/semanticTokens/refresh(2)
I[16:24:28.366] <-- reply(2)
E[16:24:28.374] IncludeCleaner: Failed to get an entry for resolved path : No such file or directory
E[16:24:28.374] IncludeCleaner: Failed to get an entry for resolved path : No such file or directory
E[16:24:28.374] IncludeCleaner: Failed to get an entry for resolved path : No such file or directory
E[16:24:28.374] IncludeCleaner: Failed to get an entry for resolved path : No such file or directory
I[16:24:28.375] --> textDocument/publishDiagnostics
I[16:24:28.375] --> textDocument/inactiveRegions
I[16:24:28.375] --> textDocument/clangd.fileStatus
I[16:24:28.378] <-- textDocument/semanticTokens/full/delta(6)
I[16:24:28.378] --> reply:textDocument/semanticTokens/full/delta(6) 0 ms
I[16:24:28.378] --> textDocument/clangd.fileStatus
I[16:24:28.575] <-- textDocument/codeAction(7)
I[16:24:28.575] --> reply:textDocument/codeAction(7) 0 ms
I[16:24:28.576] --> textDocument/clangd.fileStatus
I[16:24:28.576] <-- textDocument/foldingRange(8)
I[16:24:28.577] --> reply:textDocument/foldingRange(8) 0 ms
I[16:24:29.159] --> $/progress
I[16:24:29.160] --> $/progress
I[16:24:29.160] --> $/progress
I[16:24:29.513] Indexed src/cmds/timeline_cmd.cpp (10872 symbols, 34242 refs, 191 files)
I[16:24:29.513] Failed to compile /localdisk1/rdong3/cuda_atomic_utilization/src/cmds/timeline_cmd.cpp, index may be incomplete
I[16:24:29.571] --> $/progress
I[16:25:10.013] <-- textDocument/hover(9)
I[16:25:10.014] --> reply:textDocument/hover(9) 1 ms
I[16:25:10.014] --> textDocument/clangd.fileStatus
I[16:25:10.992] <-- textDocument/codeAction(10)
I[16:25:10.992] --> reply:textDocument/codeAction(10) 0 ms
I[16:25:10.992] --> textDocument/clangd.fileStatus
I[16:25:25.819] <-- textDocument/codeAction(11)
I[16:25:25.820] --> reply:textDocument/codeAction(11) 0 ms
I[16:25:25.820] --> textDocument/clangd.fileStatus
I[16:25:27.215] <-- textDocument/didClose
I[16:25:27.215] --> textDocument/publishDiagnostics
I[16:25:27.242] <-- textDocument/didOpen
I[16:25:27.243] <-- textDocument/documentSymbol(12)
I[16:25:27.243] <-- textDocument/codeAction(13)
I[16:25:27.243] ASTWorker building file /localdisk1/rdong3/cuda_atomic_utilization/src/timeline/mixed_cas_atomic.cu version 1 with command 
[/localdisk1/rdong3/cuda_atomic_utilization]
/usr/local/cuda-12.1/bin/nvcc -c -Xcompiler -fPIE -Ithirdparty -Isrc -I/usr/local/cuda/include -I/localdisk1/rdong3/cuda_atomic_utilization/build/.packages/a/apache-arrow/latest/4e0143c97b65425b855ad5fd03038b6a/include --dopt=on -m64 -rdc=true -gencode -gencode -gencode -resource-dir=/localdisk1/rdong3/vscode-server/data/User/globalStorage/llvm-vs-code-extensions.vscode-clangd/install/17.0.3/clangd_17.0.3/lib/clang/17 -- /localdisk1/rdong3/cuda_atomic_utilization/src/timeline/mixed_cas_atomic.cu
I[16:25:27.246] --> textDocument/clangd.fileStatus
I[16:25:27.260] <-- textDocument/semanticTokens/full(14)
I[16:25:27.260] <-- textDocument/documentLink(15)
I[16:25:27.482] Built preamble of size 4347736 for file /localdisk1/rdong3/cuda_atomic_utilization/src/timeline/mixed_cas_atomic.cu version 1 in 0.24 seconds
I[16:25:27.483] --> workspace/semanticTokens/refresh(3)
I[16:25:27.483] <-- reply(3)
E[16:25:27.517] IncludeCleaner: Failed to get an entry for resolved path : No such file or directory
I[16:25:27.519] --> textDocument/publishDiagnostics
I[16:25:27.519] --> textDocument/inactiveRegions
I[16:25:27.520] --> reply:textDocument/documentSymbol(12) 277 ms
I[16:25:27.520] --> reply:textDocument/codeAction(13) 277 ms
I[16:25:27.520] --> reply:textDocument/semanticTokens/full(14) 260 ms
I[16:25:27.520] --> reply:textDocument/documentLink(15) 260 ms
I[16:25:27.520] --> textDocument/clangd.fileStatus
I[16:25:27.650] <-- textDocument/inlayHint(16)
I[16:25:27.650] <-- textDocument/foldingRange(17)
I[16:25:27.650] <-- textDocument/codeAction(18)
I[16:25:27.650] --> reply:textDocument/inlayHint(16) 0 ms
I[16:25:27.650] --> reply:textDocument/codeAction(18) 0 ms
I[16:25:27.650] --> textDocument/clangd.fileStatus
I[16:25:27.651] --> reply:textDocument/foldingRange(17) 1 ms
I[16:25:28.021] <-- textDocument/hover(19)
I[16:25:28.021] --> reply:textDocument/hover(19) 0 ms
I[16:25:28.021] --> textDocument/clangd.fileStatus
I[16:25:28.023] <-- textDocument/codeAction(20)
I[16:25:28.023] --> reply:textDocument/codeAction(20) 0 ms
I[16:25:28.023] --> textDocument/clangd.fileStatus
I[16:25:28.179] <-- textDocument/semanticTokens/full/delta(21)
I[16:25:28.180] --> reply:textDocument/semanticTokens/full/delta(21) 0 ms
I[16:25:28.180] --> textDocument/clangd.fileStatus
```

While the compiler is gcc, clang-specific arguments like `-resource-dir` and `--` separator are generated, causing the compilation to fail. A quick test with these arguments removed showed that gcc can precompile the headers correctly.

**System information**

Output of `clangd --version`:

```
clangd version 17.0.3 (https://github.com/llvm/llvm-project 888437e1b60011b8a375dd30928ec925b448da57)
Features: linux+grpc
Platform: x86_64-unknown-linux-gnu
```

Editor/LSP plugin: VSCode + clangd

Operating system: Ubuntu 22.04

**Comment by HighCommander4** [comment]
> While the compiler is gcc, clang-specific arguments like `-resource-dir` and `--` separator are generated, causing the compilation to fail.

The log output is a bit misleading in this respect: clang is not actually invoking `gcc` with those arguments. The arguments are being interpreted by **clang** (as if you ran the command with the argv[0] replaced with `clang`, except it happens inside the `clangd` process without spawning a new `clang` process). The original value of argv[0] (e.g. `/usr/bin/gcc`) is only used to inform where clang will look for system headers.

So, the presence of these flags is not the reason for the compilation failing. To investigate the actual reason further, please provide logs with `--log=verbose`, and if you're able to, a code example.

**Comment by rongcuid** [comment]
I've got the following simple project

main.cu:

```
#include "device_launch_parameters.h"

#include <stdio.h>
#include <cstdint>
#include "test.hpp"

cudaError_t addWithCuda(int *c, const int *a, const int *b, unsigned int size);

__global__ void addKernel(int *c, const int *a, const int *b)
{
    int i = threadIdx.x;
    c[i] = a[i] + b[i];
}

int main()
{
    A aaa;
    const int arraySize = 5;
    const int a[arraySize] = {1, 2, 3, 4, 5};
    const int b[arraySize] = {10, 20, 30, 40, 50};
    int c[arraySize] = {0};

    // Add vectors in parallel.
    cudaError_t cudaStatus = addWithCuda(c, a, b, arraySize);
    if (cudaStatus != cudaSuccess)
    {
        fprintf(stderr, "addWithCuda failed!");
        return 1;
    }

    printf("{1,2,3,4,5} + {10,20,30,40,50} = {%d,%d,%d,%d,%d}\n",
           c[0], c[1], c[2], c[3], c[4]);

    // cudaDeviceReset must be called before exiting in order for profiling and
    // tracing tools such as Nsight and Visual Profiler to show complete traces.
    cudaStatus = cudaDeviceReset();
    if (cudaStatus != cudaSuccess)
    {
        fprintf(stderr, "cudaDeviceReset failed!");
        return 1;
    }

    return 0;
}

// Helper function for using CUDA to add vectors in parallel.
cudaError_t addWithCuda(int *c, const int *a, const int *b, unsigned int size)
{
    int *dev_a = 0;
    int *dev_b = 0;
    int *dev_c = 0;
    cudaError_t cudaStatus;

    // Choose which GPU to run on, change this on a multi-GPU system.
    cudaStatus = cudaSetDevice(0);
    if (cudaStatus != cudaSuccess)
    {
        fprintf(stderr, "cudaSetDevice failed!  Do you have a CUDA-capable GPU installed?");
        goto Error;
    }

    // Allocate GPU buffers for three vectors (two input, one output)    .
    cudaStatus = cudaMalloc((void **)&dev_c, size * sizeof(int));
    if (cudaStatus != cudaSuccess)
    {
        fprintf(stderr, "cudaMalloc failed!");
        goto Error;
    }

    cudaStatus = cudaMalloc((void **)&dev_a, size * sizeof(int));
    if (cudaStatus != cudaSuccess)
    {
        fprintf(stderr, "cudaMalloc failed!");
        goto Error;
    }

    cudaStatus = cudaMalloc((void **)&dev_b, size * sizeof(int));
    if (cudaStatus != cudaSuccess)
    {
        fprintf(stderr, "cudaMalloc failed!");
        goto Error;
    }

    // Copy input vectors from host memory to GPU buffers.
    cudaStatus = cudaMemcpy(dev_a, a, size * sizeof(int), cudaMemcpyHostToDevice);
    if (cudaStatus != cudaSuccess)
    {
        fprintf(stderr, "cudaMemcpy failed!");
        goto Error;
    }

    cudaStatus = cudaMemcpy(dev_b, b, size * sizeof(int), cudaMemcpyHostToDevice);
    if (cudaStatus != cudaSuccess)
    {
        fprintf(stderr, "cudaMemcpy failed!");
        goto Error;
    }

    // Launch a kernel on the GPU with one thread for each element.
    addKernel<<<1, size>>>(dev_c, dev_a, dev_b);

    // Check for any errors launching the kernel
    cudaStatus = cudaGetLastError();
    if (cudaStatus != cudaSuccess)
    {
        fprintf(stderr, "addKernel launch failed: %s\n", cudaGetErrorString(cudaStatus));
        goto Error;
    }

    // cudaDeviceSynchronize waits for the kernel to finish, and returns
    // any errors encountered during the launch.
    cudaStatus = cudaDeviceSynchronize();
    if (cudaStatus != cudaSuccess)
    {
        fprintf(stderr, "cudaDeviceSynchronize returned error code %d after launching addKernel!\n", cudaStatus);
        goto Error;
    }

    // Copy output vector from GPU buffer to host memory.
    cudaStatus = cudaMemcpy(c, dev_c, size * sizeof(int), cudaMemcpyDeviceToHost);
    if (cudaStatus != cudaSuccess)
    {
        fprintf(stderr, "cudaMemcpy failed!");
        goto Error;
    }

Error:
    cudaFree(dev_c);
    cudaFree(dev_a);
    cudaFree(dev_b);

    return cudaStatus;
}
```

test.hpp:
```
#pragma once

#include <cstdint>

class A {};
```

There are two different errors.

`main.cu` generates this, despite not having sm_35 as argument (maybe a separate issue):

```
Unknown argument: '-rdc=true'clang(drv_unknown_argument)
Unknown argument: '-gencode'clang(drv_unknown_argument)
GPU arch sm_35 is supported by CUDA versions between 7.0 and 11.8 (inclusive), but installation at /usr/local/cuda is 12.1; use '--cuda-path' to specify a different CUDA install, pass a different GPU arch with '--cuda-gpu-arch', or pass '--no-cuda-version-check'clang(drv_cuda_version_unsupported)
device_launch_parameters.h
/usr/local/cuda-12.1/targets/x86_64-linux/include/device_launch_parameters.h
```

`test.hpp` generates this:

```
Unknown argument: '-rdc=true'clang(drv_unknown_argument)
Unknown argument: '-gencode'clang(drv_unknown_argument)
'cstdint' file not foundclang(pp_file_not_found)
```

The following are `compile_commads.json` generated by my build system (xmake) and the clangd log:

[clangd.log](https://github.com/clangd/clangd/files/13942378/clangd.log)
[compile_commands.json](https://github.com/clangd/clangd/files/13942379/compile_commands.json)

**Comment by HighCommander4** [comment]
> ```
> Unknown argument: '-rdc=true'clang(drv_unknown_argument)
> Unknown argument: '-gencode'clang(drv_unknown_argument)
> ```

In some other issues filed about CUDA (https://github.com/clangd/clangd/issues/1610, https://github.com/clangd/clangd/discussions/1863), I have seen people remove those flags using clangd's config file.

> 'cstdint' file not foundclang(pp_file_not_found)

From the log, it looks like clangd is looking for it in `/usr/lib/gcc/x86_64-linux-gnu/12/../../../../include/c++`, does this directory exist?

**Comment by rongcuid** [comment]
> > ```
> > Unknown argument: '-rdc=true'clang(drv_unknown_argument)
> > Unknown argument: '-gencode'clang(drv_unknown_argument)
> > ```
> 
> In some other issues filed about CUDA (#1610, #1863), I have seen people remove those flags using clangd's config file.
> 
> > 'cstdint' file not foundclang(pp_file_not_found)
> 
> From the log, it looks like clangd is looking for it in `/usr/lib/gcc/x86_64-linux-gnu/12/../../../../include/c++`, does this directory exist?

The real C++ directory seems to be under `/usr/lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/11`. or more shortly `/usr/include/c++/11`

**Comment by HighCommander4** [comment]
> The real C++ directory seems to be under `/usr/lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/11`. or more shortly `/usr/include/c++/11`

This may be a case of https://github.com/clangd/clangd/issues/1394 then, see workarounds in [this comment](https://github.com/clangd/clangd/issues/1394#issuecomment-1328676884).
